sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -465,19 +519,33 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
           val alwaysTrueRef = AttributeReference(ALWAYS_TRUE_COLNAME,
             BooleanType)(exprId = alwaysTrueExprId)
 
-          val aggValRef = query.output.head
-
           if (havingNode.isEmpty) {
             // CASE 2: Subquery with no HAVING clause
+
+            // We replace original expression id with a new one. The added Alias column
+            // must use expr id of original output. If we don't replace old expr id in the
+            // query, the added Project in potential Project-Filter-Project can be removed
+            // by removeProjectBeforeFilter in ColumnPruning.
```

Comment:
shall we fix `removeProjectBeforeFilter` to only remove attribute-only projects?

Reply:
Worth trying, right now not sure if any other thing will be affected.

Reply:
Tried locally. Added subquery tests are passed. We can see if Jenkins passes.

Reply:
Ok. Seems fine. Jenkins passes.